core/src/main/java/org/eclipse/hono/util/EncodedPassword.java

```diff
@@ -0,0 +1,170 @@
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     Red Hat - initial creation
+ *******************************************************************************/
+package org.eclipse.hono.util;
+
+import io.vertx.core.json.JsonObject;
+import org.springframework.security.crypto.password.PasswordEncoder;
+
+import java.util.Base64;
+import java.util.Objects;
+
+/**
+ * Helper class for parsing and encoding passwords.
+ * <p>
+ * The encoded password have the following main elements:
+ * <ul>
+ *     <li><em>Hash function</em> that are used to encode the password</li>
+ *     <li><em>Salt</em> which is optionally used for encoding</li>
+ *     <li><em>Raw password</em> we want to encode</li>
+ *     <li><em>Password Hash</em> Hashed value of the password</li>
+ * </ul>
+ * <p>
+ */
+public class EncodedPassword {
+    private static final String PREFIX = "{";
+    private static final String SUFFIX = "}";
+
+    byte[] salt;
+    String rawPassword;
+    String passwordHash;
+    String hashFunction =  CredentialsConstants.DEFAULT_HASH_FUNCTION;
+
+    /**
+     * Creates a new instance from Hono Secret.
+     *
+     * The secret is expected to be of type <em>hashed-password</em> as defined by
+     * <a href="https://www.eclipse.org/hono/api/credentials-api/">Hono's Credentials API</a>.
+     *
+     * @param honoSecret JSON object that contains hono-formatted secret
+     * @return Encoded password
+     * @throws IllegalArgumentException if candidate hashed-password secret does not contain a password hash
+     */
+    public static EncodedPassword fromHonoSecret(final JsonObject honoSecret) throws IllegalArgumentException {
+        Objects.requireNonNull(honoSecret);
+
+        final String pwdHash = honoSecret.getString(CredentialsConstants.FIELD_SECRETS_PWD_HASH);
+        if (pwdHash == null) {
+            throw new IllegalArgumentException("candidate hashed-password secret does not contain a pwd hash");
+        }
+
+        final String hashFunction = honoSecret.getString(
+                CredentialsConstants.FIELD_SECRETS_HASH_FUNCTION,
+                CredentialsConstants.DEFAULT_HASH_FUNCTION);
+
+        final String encodedSalt = honoSecret.getString(CredentialsConstants.FIELD_SECRETS_SALT);
+        final EncodedPassword encodedPassword = new EncodedPassword();
+        encodedPassword.hashFunction = hashFunction;
+        encodedPassword.passwordHash = pwdHash;
+        if (encodedSalt != null) {
+            encodedPassword.salt = Base64.getDecoder().decode(encodedSalt);
+        }
+        return encodedPassword;
+    }
+
+    /**
+     * Creates a new instance from the String in {salt}password format. The password can be in raw or hashed form.
+     * This method is used by Spring Security password encoders, to parse submitted values.
+     *
+     * @param formattedPassword Password in the {salt}password format
+     * @param raw {@true} if password is in raw form, {@false} if it's hashed
```

Comment:
I do not understand what *raw* vs *hashed* means. Examples in the JavaDoc would *really* be helpful to understand what this method actually does. Apart from that I strongly believe that using a parameter to modify how another parameter needs to be interpreted is very bad practice and makes it very hard to both document and understand the method. I would therefore suggest to split this up into two dedicated methods whose purpose can then be easily described. WDYT?